cmd/snap-seccomp-blacklist/snap-seccomp-blacklist.c

```diff
@@ -0,0 +1,215 @@
+     * correct operation was confirmed using the pseudo-code program and the
+     * bpf_dbg tool from the kernel tools/bpf directory.
+     *
+     * NOTE: I'm not using scmp_rule_add_exact as that was not doing anything
+     * at all (presumably due to having all the architectures defined). */
+
+    const struct scmp_arg_cmp no_tty_inject = {
+        /* When faced with ioctl(fd, request); refuse to proceed when
+         * request&0xffffffff == TIOCSTI. This specific way to encode the
+         * filter has the following important properties:
+         *
+         * 1) it blocks ioctl(fd, TIOCSTI, ptr).
+         * 1) it also blocks ioctl(fd, (1UL<<32) | TIOCSTI, ptr).
+         * 2) it doesn't block ioctl(fd, (1UL<<32) | (request not equal to TIOCSTI), ptr);
+         *
+         * We learned that existing programs make legitimate requests with all
+         * bits set in the more significant 32bit word of the 64 bit double
+         * word. While this kernel behavior remains suspect and presumably
+         * undesired it is unlikely to change for backwards compatibility
+         * reasons. As such we cannot block all requests with high-bits set. */
+        .arg = 1,
+        .op = SCMP_CMP_MASKED_EQ,
+        .datum_a = 0xffffffffUL,
+        .datum_b = TIOCSTI,
+    };
+    sc_err = seccomp_rule_add(ctx, SCMP_ACT_ERRNO(EPERM), sys_ioctl_nr, 1, no_tty_inject);
+
+    if (sc_err < 0) {
+        showerr("cannot add rule preventing the use high bits in ioctl");
+        return sc_err;
+    }
+    return 0;
+}
+
+typedef struct arch_set {
+    const char *name;
+    const uint32_t *arch_tags;
+    size_t num_arch_tags;
+} arch_set;
+
+int main(int argc, char **argv) {
+    const uint32_t le_arch_tags[] = {
+        SCMP_ARCH_X86_64,
+        SCMP_ARCH_AARCH64,
+        SCMP_ARCH_PPC64LE,
+        SCMP_ARCH_S390X,
+    };
+    const uint32_t be_arch_tags[] = {
+        SCMP_ARCH_S390X,
+    };
+    const arch_set arch_sets[] = {
+        {"LE", le_arch_tags, sizeof le_arch_tags / sizeof *le_arch_tags},
+        {"BE", be_arch_tags, sizeof be_arch_tags / sizeof *be_arch_tags},
+    };
+    int rc = -1;
+
+    for (size_t i = 0; i < sizeof arch_sets / sizeof *arch_sets; ++i) {
+        const arch_set *as = &arch_sets[i];
+        int sc_err;
+        int fd = -1;
+        int fname_len;
+        char fname[PATH_MAX];
+
+        scmp_filter_ctx ctx = NULL;
+        ctx = seccomp_init(SCMP_ACT_ALLOW);
+        if (ctx == NULL) {
+            showerr("cannot construct seccomp context");
+            return -rc;
+        }
+        sc_err = populate_filter(ctx, as->arch_tags, as->num_arch_tags);
+        if (sc_err < 0) {
+            seccomp_release(ctx);
+            return -rc;
+        }
+
+        /* Save pseudo-code program */
+        fname_len = snprintf(fname, sizeof fname, "%s-blacklist.pfc", as->name);
+        if (fname_len < 0 || fname_len >= sizeof fname) {
+            showerr("cannot format file name");
```

Comment:
Should this error have more context (file name, arch...)?

Reply:
Given that this is a single-trick-pony I don't think this is required.

Reply:
I added some context to error messages here and elsewhere.